### DISCUSSION TITLE
help with audio effects and espnow receive

### DISCUSSION THREAD

**Discussion by nipper01792** [pschatzmann/arduino-audio-tools#565] [Q&A] [answered]
hi everyone I wonder if someone can help me , I want to apply an audio effect on the incoming decoded signal coming from espnow . I cannot seem to get this to work . 
my constructor at the moment is
ESPNowStream now;
AudioKitStream kit; // Access I2S as stream

VolumeStream volume(kit);

AudioEffects<GeneratorFromStream<effect_t>> effects(volume,1,1.0); // apply effects on kit
GeneratedSoundStream<int16_t> fx(effects);

 EncodedAudioStream decoder(&fx, new SBCDecoder(1256)); // decode and write to I2S - ESP Now is limited to 256 bytes
StreamCopy copier(decoder,now);     

I have tried multiple configurations of this but following the logic this should work, 
it does compile but I get 
[E] AudioStreams.h : 86 - AudioStreamX: unsupported operation!
in the serial . please help

**Comment by nipper01792** [comment]
I think I've confused myself as I've tried to fix this so many times, I'm lost

**Comment by nipper01792** [comment]
ok so lets start again .
ive started with this 
ESPNowStream now;
AudioKitStream kit; // Access I2S as stream
VolumeStream volume(kit);
 EncodedAudioStream decoder(&volume, new SBCDecoder(1256)); 
StreamCopy copier(decoder,now);   

where should I put the effects in the chain.

**Comment by pschatzmann** [comment]
I don't think that currently there is a clean way to implement this. Maybe it might work if you create two separate chains that are connected via a QueueStream. It will be output of the decoder and input for the effects.

I will investigate if I can extend the framework somehow to support this in a better way

**Reply by nipper01792** [reply]
ok now I understand. I wondered why I got so confused, I know my original chain was wrong but it was the only way I could compile. thank you for you help

**Reply by nipper01792** [reply]
also I have tried this on the espnow send side, it works 
 but I couldn't fix the stutter hence why I was trying to do this on the receive side

**Comment by pschatzmann** [comment]
I am providing now a proper Stream implementation for the effects with https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_audio_effect_stream_t.html which should be easier to use.

It can be used both, on the input and output side. So this should work in your case as well.

I also updated the [documentation](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects).

**Reply by nipper01792** [reply]
thank you so much, you've gone above and beyond for me , you are a superstar. thank you . I will try this asap .

**Comment by nipper01792** [comment]
mmm doesn't seem to work with espnow send or receive.

here's the send chain

ESPNowStream now;
AudioKitStream kit; 
AudioEffectStream effects(kit);
EncodedAudioStream encoder(&now, new SBCEncoder());
StreamCopy copier(encoder,effects);

**Comment by nipper01792** [comment]
it kills the input/output completely so no audio data is send via espnow

**Comment by pschatzmann** [comment]
Are you sure that you called begin on AudioEffectStream ?

**Comment by nipper01792** [comment]
yes the sketch works perfectly if take out  the effects from the chain

**Comment by nipper01792** [comment]
#include "AudioTools.h"
#include "AudioLibs/Communication.h"
#include "AudioCodecs/CodecSBC.h"
#include "AudioLibs/AudioKit.h"
//uint16_t sample_rate = 44100;
//uint8_t channels = 1;  // The stream will have 2 channels

#define AI_THINKER_ES8388_VOLUME_HACK

const int sample_rate = 44100;
const int channels = 1;
const int bits_per_sample = 16;

// Effects control input
float volumeControl = 0;//0-1

// Effects
Boost boost(volumeControl);






ESPNowStream now;
AudioKitStream kit; // Access I2S as stream
AudioEffectStream effects(kit);

EncodedAudioStream encoder(&now, new SBCEncoder()); // encode and write to ESP-now
StreamCopy copier(encoder,kit);  // copies sound into i2s
const char *peers[] = {"A8:48:FA:0B:93:01"};
void updateValues(){
  // update values in controls
  boost.setVolume(volumeControl);
  boost.setActive(volumeControl>0);
  
 
}
void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  
  
  // setup effects
  effects.addEffect(boost);
  
  
  updateValues();
  
  auto cfg = now.defaultConfig();
  cfg.mac_address = "A8:48:FA:0B:93:02";
  now.begin(cfg);
  now.addPeers(peers);

 
   auto cfg1 = kit.defaultConfig(RX_MODE);
cfg1.i2s_format = I2S_LSB_FORMAT; 
  cfg1.bits_per_sample = 16;
  cfg1.channels = channels;
//cfg1.buffer_count = 2;
//cfg1.buffer_size =256;
  cfg1.sample_rate = 44100;
  cfg1.input_device = AUDIO_HAL_ADC_INPUT_LINE2; // microphone
  kit.begin(cfg1);
 kit.setVolume(1.0);

  // start encoder
  encoder.begin(cfg1);
  
  Serial.println("Sender started...");
}

void loop() { 
  copier.copy();
}


its the same settings  at the receiving end

**Comment by pschatzmann** [comment, marked as answer]
You do not call begin on AudioEffectStream !

**Comment by nipper01792** [comment]
how the ...... I'm so far down the rabbit hole I didn't even realise id deleted that... so sorry to waste your time, yes I can confirm that it it works great . thank you so much . ill send some coffee
